src/dmd/expressionsem.d

```diff
@@ -5893,36 +5851,56 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                 auto tiargs = new Objects(3);
                 Loc loc = exp.e1.loc;
 
+                /*
+                Stores the result of an operand expression into a temporary
+                if necessary, e.g. if it is an impure fuction call containing side
+                effects as in https://issues.dlang.org/show_bug.cgi?id=20114
+
+                Params:
+                    op = an expression which may require a temporary and will be
+                         replaced by `(auto tmp = op, tmp)` if necessary
+
+                Returns: `op` or `tmp` for subsequent access to the possibly promoted operand
+                */
+                Expression maybePromoteToTmp(ref Expression op)
```

Comment:
This function should remain at it's previous location s.t. it can be used for other expressions.
(I would like to implement unary expressions as well, e.g. `assert(!i)`)

Reply:
Okay thx for rationale, will move it back.